### PR TITLE
Add Strumpack

### PR DESCRIPTION
**Pull request by jamiebramwell**

### CODE REVIEW

File: host-configs/quartz-toss_4_x86_64_ib-gcc@10.3.1.cmake [hunk 1]
```diff
@@ -97,7 +99,7 @@ set(SUNDIALS_DIR "${TPL_ROOT}/sundials-6.5.1-3zqikxu2ss427k5jvnl6z2rihmt2mwjf" C
 
 set(UMPIRE_DIR "${TPL_ROOT}/umpire-2022.10.0-gayss6vwjqfnw76p5cvti5lvhvfprzar" CACHE PATH "")
 
-set(TRIBOL_DIR "${TPL_ROOT}/tribol-0.1.0.6-xflzmbinmj5eycoe3qnoyebwlyjxirqm" CACHE PATH "")
+# TRIBOL not built
```
Comment (chapman39) [inline host-configs/quartz-toss_4_x86_64_ib-gcc@10.3.1.cmake]: should adiak, caliper, tribol be built?

**Comment by chapman39** [comment]
@jamiebramwell I believe quartz jobs are failing since the new TPL roots needs group access `chmod -R g+rwX`

**Comment by jamiebramwell** [comment]
Sigh. I was going to look at that tonight. Hopefully it will be done at the end of today! Thanks @chapman39 !

File: scripts/spack/configs/docker/ubuntu20/spack.yaml [hunk 2]
```diff
@@ -9,7 +9,7 @@ spack:
     build_stage::
     - $spack/../build_stage
 
-  compilers::
```
Comment (chapman39) [inline scripts/spack/configs/docker/ubuntu20/spack.yaml]: Unless this was causing an issue we should still have the double colon `::`. It's a spack thing - it pretty much means "only use the compilers below".

**Review by chapman39 — APPROVED** [review]
im not a very helpful reviewer for `src`. but in terms of spack/ build system stuff lgtm!

File: scripts/spack/configs/docker/ubuntu20/spack.yaml [hunk 3]
```diff
@@ -9,7 +9,7 @@ spack:
...
+  compilers:
```
Comment (chapman39) [inline scripts/spack/configs/docker/ubuntu20/spack.yaml]:
```suggestion
  compilers::
```

Comment (jamiebramwell) [reply scripts/spack/configs/docker/ubuntu20/spack.yaml]: I did not know this! Thank you for noticing!

File: CMakeLists.txt [hunk 4]
```diff
@@ -43,6 +43,10 @@ if(NOT DEFINED SERAC_SOURCE_DIR)
 endif()
 message(STATUS "Serac Source Dir: ${SERAC_SOURCE_DIR}")
 
+if(SERAC_ENABLE_CODEVELOP)
```
Comment (white238) [inline CMakeLists.txt:46]: This saddens me deeply.

File: cmake/thirdparty/SetupSeracThirdParty.cmake [hunk 5]
```diff
@@ -221,6 +221,15 @@ if (NOT SERAC_THIRD_PARTY_LIBRARIES_FOUND)
             set(SuperLUDist_DIR ${SUPERLUDIST_DIR} CACHE PATH "")
             set(MFEM_USE_SUPERLU ${ENABLE_MPI} CACHE BOOL "")
         endif()
+        if(STRUMPACK_DIR)
+            serac_assert_is_directory(VARIABLE_NAME STRUMPACK_DIR)
+            set(MFEM_USE_STRUMPACK ON CACHE BOOL "")
+            find_package(strumpack CONFIG PATHS ${STRUMPACK_DIR}/lib/cmake/STRUMPACK)
```
Comment (white238) [inline cmake/thirdparty/SetupSeracThirdParty.cmake:227]: I've forgotten, did we end up confirming we needed this `find_package`?

File: scripts/docker/dockerfile_gcc-11 [hunk 6]
```diff
@@ -16,7 +16,7 @@ RUN git clone --recursive --branch $branch --single-branch --depth 1 https://git
 # Build/install TPLs via spack and then remove the temporary build directory on success
 RUN cd serac_repo && python3 ./scripts/uberenv/uberenv.py --spack-env-file=./scripts/spack/configs/docker/ubuntu20/spack.yaml \
                                                           --project-json=.uberenv_config.json \
-                                                          --spec=%gcc@11.1.0 --prefix=/home/serac/serac_tpls -k \
+                                                          --spec=%gcc@11.1.0~strumpack --prefix=/home/serac/serac_tpls -k \
```
Comment (white238) [inline scripts/docker/dockerfile_gcc-11:19]: This doesn't work anymore either?

File: scripts/spack/packages/serac/package.py [hunk 7]
```diff
@@ -139,6 +142,9 @@ class Serac(CachedCMakePackage, CudaPackage):
 
     depends_on("superlu-dist@8.1.2")
 
+    # The optional slate dependency is not handled in the MFEM spack package
+    depends_on("strumpack~slate~butterflypack~shared", when="+strumpack")
```
Comment (white238) [inline scripts/spack/packages/serac/package.py]: You should remove the `~shared` and add `strumpack` to the for loop on line 150.

File: src/serac/numerics/equation_solver.hpp [hunk 8]
```diff
@@ -192,6 +192,62 @@ class SuperLUSolver : public mfem::Solver {
...
+    }
+  }
+
+  /**
+   * @brief Factor and solve the linear system y = Op^{-1} x using Strumpack
+   *
+   * @param x The input RHS vector
+   * @param y The output solution vector
+   */
+  void Mult(const mfem::Vector& x, mfem::Vector& y) const;
```
Comment (white238) [inline src/serac/numerics/equation_solver.hpp]: Should these be called `input` and `output`?

File: cmake/thirdparty/SetupSeracThirdParty.cmake (hunk 5 above)
Comment (jamiebramwell) [reply cmake/thirdparty/SetupSeracThirdParty.cmake:227]: unsure - I can check

File: scripts/docker/dockerfile_gcc-11 (hunk 6 above)
Comment (jamiebramwell) [reply scripts/docker/dockerfile_gcc-11:19]: no - it was the same weird problem. It showed up on my machine when I had docker installed - so I think it has something to do with that.

File: src/serac/numerics/equation_solver.hpp (hunk 8 above)
Comment (jamiebramwell) [reply src/serac/numerics/equation_solver.hpp]: Yes - this is an mfem-ism I picked up.

File: cmake/thirdparty/SetupSeracThirdParty.cmake (hunk 5 above)
Comment (jamiebramwell) [reply cmake/thirdparty/SetupSeracThirdParty.cmake:227]: yeah, we need it.

**Comment by jamiebramwell** [comment]
I believe this is done now.

**Review by white238 — APPROVED** [review]

**Pull request merged**